services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/validation/HankeValidator.kt

```diff
@@ -15,78 +17,112 @@ class HankeValidator : ConstraintValidator<ValidHanke, Hanke> {
-            ok1 = checkMandatoryYhteystietoData(yhteystieto, context, ok1)
+            // mandatory
+            ok = checkMandatoryYhteystietoData(yhteystieto, context, ok)
         }
         hanke.arvioijat.forEach { yhteystieto ->
-            //mandatory
-            ok1 = checkMandatoryYhteystietoData(yhteystieto, context, ok1)
+            // mandatory
+            ok = checkMandatoryYhteystietoData(yhteystieto, context, ok)
         }
-        return ok1
+        return ok
     }
 
     private fun checkMandatoryYhteystietoData(yhteystieto: HankeYhteystieto, context: ConstraintValidatorContext, ok: Boolean): Boolean {
         var ok1 = ok
-        if (!yhteystieto.sukunimi.isNullOrBlank() || !yhteystieto.etunimi.isNullOrBlank()
-                || !yhteystieto.email.isNullOrBlank() || !yhteystieto.puhelinnumero.isNullOrBlank()) {
-            //if any of the attributes contains something then all must exist
-            if(yhteystieto.sukunimi.isNullOrBlank() || yhteystieto.etunimi.isNullOrBlank()
-                    || yhteystieto.email.isNullOrBlank() || yhteystieto.puhelinnumero.isNullOrBlank()) {
+        if (!yhteystieto.sukunimi.isBlank() || !yhteystieto.etunimi.isBlank()
+                || !yhteystieto.email.isBlank() || !yhteystieto.puhelinnumero.isBlank()) {
+            // if any of the attributes contains something then all must exist
+            if (yhteystieto.sukunimi.isBlank() || yhteystieto.etunimi.isBlank()
+                    || yhteystieto.email.isBlank() || yhteystieto.puhelinnumero.isBlank()) {
+                // TODO: is that property node correct?
+                // TODO: Does not currently matter, though, as the node information does not get through to error response.
                 context.buildConstraintViolationWithTemplate(HankeError.HAI1002.toString()).addPropertyNode("YhteysTiedot").addConstraintViolation()
                 ok1 = false
             }
         }
         return ok1
     }
-}
\ No newline at end of file
+
+    private fun checkTyomaaTiedot(hanke: Hanke, context: ConstraintValidatorContext): Boolean {
+        var ok = true
+        // tyomaaKatuosoite - either null or length <= maximum
+        if (hanke.tyomaaKatuosoite != null && hanke.tyomaaKatuosoite!!.length > MAXIMUM_TYOMAAKATUOSOITE_LENGTH) {
+            context.buildConstraintViolationWithTemplate(HankeError.HAI1002.toString()).addPropertyNode("tyomaaKatuosoite").addConstraintViolation()
+            ok = false
+        }
+
+        return ok
+    }
+
+    private fun checkHaitat(hanke: Hanke, context: ConstraintValidatorContext): Boolean {
+        var ok = true
+        // TODO: can haitta start/end after the hanke ends? E.g. if agreed that another hanke will continue with the same hole soon after?
+        // haittaAlkuPvm - either null or after alkuPvm and before maximum end date
+        if (hanke.haittaAlkuPvm != null && (hanke.haittaAlkuPvm!!.isBefore(hanke.alkuPvm) || hanke.haittaAlkuPvm!!.isAfter(MAXIMUM_DATE))) {
```

Comment:
See below:` && (hanke.haittaLoppuPvm!!.isBefore(hanke.haittaAlkuPvm) || `

Reply:
eh, I read that wrong.. so, "nevermind" :P
I added those limits merely by thinking what might be possible and what not; I'd think that if something about the hanke is already causing problems, then the hanke must have also have started earlier.  (I.e. if the city is talking that hanke can not start before the same day, then the problems related to it can not start earlier either..)